# WEEK05
## 数据透视表
### 定义
* 数据透视表（Pivot Table）是一种交互式的表，可以进行某些计算，如求和与计数等。所进行的计算与数据跟数据透视表中的排列有关。之所以称为数据透视表，是因为可以动态地改变它们的版面布置，以便按照不同方式分析数据，也可以重新安排行号、列标和页字段。每一次改变版面布置时，数据透视表会立即按照新的布置重新计算数据。另外，如果原始数据发生更改，则可以更新数据透视表。
### 步骤
1. 挑两个变量通常是类别变量，作为行跟列，做一个比较宽的报表
2. 筛选——抽某个变量可能有重复 表格的长和宽可以控制，控制在表格的独特值或想要部分
3. 排序
* tip：只要是数据没有量的至少可以计量（数数）
## pandas和excel做数据报表的优缺点
* excel优：可视化操作容易直接拖拉 数据量不大
    缺：更新数据麻烦成本高
* pandas优：可以做出工作流程  数据量大复杂   
    缺：没有可视化操作

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv ("20春_pandas_week02_hurun_unicorn.tsv", encoding = "utf8", sep="\t")
# df.info() # 注意看Dtype! 
分进合击初阶 = df.groupby(by= ["行业", "城市"]) \
                .agg({"估值（亿人民币）":"sum"}) 
分进合击初阶

估值（亿人民币）
行业   城市                                
3D印刷 Burlington Massachussets       150
     萨默维尔市                           70
     雷德伍德城                          150
云计算  -                               70
     上海                              70
...                                 ...
金融科技 芝加哥                            220
     诺伊达                            700
     门洛帕克                           550
     首尔                              70
     香港                             140

[298 rows x 1 columns]

* 注意是有298列，1行数据
* 如果想把表格变宽，城市作为行，有2种方法

In [6]:
# 第一种 groupby+agg+pivot
分进合击迸阶pv = df.groupby(by= ["行业","城市"]) \
                  .agg({"估值（亿人民币）":"sum"}) \
                  .reset_index() \
                  .set_index(["行业"]) \
                  .pivot(columns="城市", values="估值（亿人民币）")
print (分进合击迸阶pv.shape)
分进合击迸阶pv

(25, 120)


城市           -  Burlington Massachussets  Emerville  Foster City  Guilford  \
行业                                                                           
3D印刷       NaN                     150.0        NaN          NaN       NaN   
云计算       70.0                       NaN        NaN          NaN       NaN   
人工智能       NaN                       NaN        NaN        200.0       NaN   
健康科技       NaN                       NaN        NaN          NaN      70.0   
共享经济       NaN                       NaN        NaN          NaN       NaN   
区块链      150.0                       NaN        NaN          NaN       NaN   
即时通讯       NaN                       NaN        NaN          NaN       NaN   
大数据        NaN                       NaN        NaN          NaN       NaN   
媒体和娱乐      NaN                       NaN        NaN          NaN       NaN   
房地产科技      NaN                       NaN        NaN          NaN       NaN   
教育科技       NaN                       NaN        NaN          NaN       NaN   
新能源        NaN                       NaN        NaN          NaN       NaN   
新能源汽车      NaN                       NaN        NaN          NaN       NaN   
新零售        NaN                       NaN        NaN          NaN       NaN   
机器人        NaN                       NaN        NaN          NaN       NaN   
消费品      150.0                       NaN        NaN          NaN       NaN   
游戏         NaN                       NaN        NaN          NaN       NaN   
物流         NaN                       NaN        NaN          NaN       NaN   
生命科学       NaN                       NaN        NaN          NaN       NaN   
电子商务       NaN                       NaN        NaN          NaN       NaN   
网络安全       NaN                       NaN      500.0          NaN       NaN   
航天         NaN                       NaN        NaN          NaN       NaN   
虚拟与增强现实    NaN                       NaN        NaN          NaN       NaN   
软件与服务      NaN                       NaN        NaN          NaN       NaN   
金融科技       NaN                       NaN        NaN          NaN       NaN   

城市       Harrisburg  Plantation  Stafford      上海     东京  ...  阿拉米达    阿林顿  \
行业                                                        ...                
3D印刷            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
云计算             NaN         NaN       NaN    70.0    NaN  ...   NaN    NaN   
人工智能          500.0         NaN       NaN   290.0  150.0  ...   NaN    NaN   
健康科技            NaN         NaN       NaN   970.0    NaN  ...   NaN    NaN   
共享经济            NaN         NaN       NaN   200.0    NaN  ...   NaN    NaN   
区块链             NaN         NaN       NaN     NaN   70.0  ...   NaN    NaN   
即时通讯            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
大数据             NaN         NaN       NaN    70.0    NaN  ...   NaN    NaN   
媒体和娱乐           NaN         NaN       NaN   590.0    NaN  ...   NaN    NaN   
房地产科技           NaN         NaN       NaN   170.0    NaN  ...   NaN    NaN   
教育科技            NaN         NaN       NaN   450.0    NaN  ...   NaN    NaN   
新能源             NaN         NaN      70.0    70.0    NaN  ...  70.0    NaN   
新能源汽车           NaN         NaN       NaN   600.0    NaN  ...   NaN    NaN   
新零售             NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
机器人             NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
消费品             NaN         NaN       NaN   100.0    NaN  ...   NaN    NaN   
游戏              NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
物流              NaN         NaN       NaN   440.0    NaN  ...   NaN    NaN   
生命科学            NaN         NaN       NaN   300.0    NaN  ...   NaN    NaN   
电子商务            NaN         NaN       NaN   780.0    NaN  ...   NaN    NaN   
网络安全            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
航天              NaN         NaN       NaN     NaN    NaN  ...   NaN  200.0   
虚拟与增强现实         NaN  

In [4]:
# 第二种 pivot_table
分进合击迸阶 = df.pivot_table(index= "行业", columns="城市",  \
                 values="估值（亿人民币）", aggfunc="sum") 
print (分进合击迸阶.shape)
分进合击迸阶

(25, 120)


城市           -  Burlington Massachussets  Emerville  Foster City  Guilford  \
行业                                                                           
3D印刷       NaN                     150.0        NaN          NaN       NaN   
云计算       70.0                       NaN        NaN          NaN       NaN   
人工智能       NaN                       NaN        NaN        200.0       NaN   
健康科技       NaN                       NaN        NaN          NaN      70.0   
共享经济       NaN                       NaN        NaN          NaN       NaN   
区块链      150.0                       NaN        NaN          NaN       NaN   
即时通讯       NaN                       NaN        NaN          NaN       NaN   
大数据        NaN                       NaN        NaN          NaN       NaN   
媒体和娱乐      NaN                       NaN        NaN          NaN       NaN   
房地产科技      NaN                       NaN        NaN          NaN       NaN   
教育科技       NaN                       NaN        NaN          NaN       NaN   
新能源        NaN                       NaN        NaN          NaN       NaN   
新能源汽车      NaN                       NaN        NaN          NaN       NaN   
新零售        NaN                       NaN        NaN          NaN       NaN   
机器人        NaN                       NaN        NaN          NaN       NaN   
消费品      150.0                       NaN        NaN          NaN       NaN   
游戏         NaN                       NaN        NaN          NaN       NaN   
物流         NaN                       NaN        NaN          NaN       NaN   
生命科学       NaN                       NaN        NaN          NaN       NaN   
电子商务       NaN                       NaN        NaN          NaN       NaN   
网络安全       NaN                       NaN      500.0          NaN       NaN   
航天         NaN                       NaN        NaN          NaN       NaN   
虚拟与增强现实    NaN                       NaN        NaN          NaN       NaN   
软件与服务      NaN                       NaN        NaN          NaN       NaN   
金融科技       NaN                       NaN        NaN          NaN       NaN   

城市       Harrisburg  Plantation  Stafford      上海     东京  ...  阿拉米达    阿林顿  \
行业                                                        ...                
3D印刷            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
云计算             NaN         NaN       NaN    70.0    NaN  ...   NaN    NaN   
人工智能          500.0         NaN       NaN   290.0  150.0  ...   NaN    NaN   
健康科技            NaN         NaN       NaN   970.0    NaN  ...   NaN    NaN   
共享经济            NaN         NaN       NaN   200.0    NaN  ...   NaN    NaN   
区块链             NaN         NaN       NaN     NaN   70.0  ...   NaN    NaN   
即时通讯            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
大数据             NaN         NaN       NaN    70.0    NaN  ...   NaN    NaN   
媒体和娱乐           NaN         NaN       NaN   590.0    NaN  ...   NaN    NaN   
房地产科技           NaN         NaN       NaN   170.0    NaN  ...   NaN    NaN   
教育科技            NaN         NaN       NaN   450.0    NaN  ...   NaN    NaN   
新能源             NaN         NaN      70.0    70.0    NaN  ...  70.0    NaN   
新能源汽车           NaN         NaN       NaN   600.0    NaN  ...   NaN    NaN   
新零售             NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
机器人             NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
消费品             NaN         NaN       NaN   100.0    NaN  ...   NaN    NaN   
游戏              NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
物流              NaN         NaN       NaN   440.0    NaN  ...   NaN    NaN   
生命科学            NaN         NaN       NaN   300.0    NaN  ...   NaN    NaN   
电子商务            NaN         NaN       NaN   780.0    NaN  ...   NaN    NaN   
网络安全            NaN         NaN       NaN     NaN    NaN  ...   NaN    NaN   
航天              NaN         NaN       NaN     NaN    NaN  ...   NaN  200.0   
虚拟与增强现实         NaN  

* 以上两种方法都让表格变成25列，120行数据
* 第一种方法.reset_index()去掉索引，.set_index重新加索引，pivot index前面已经做了就可以省略，步骤多但中间可以细微调整
* 第二种方法一步到位 pivot 指定列index、行columns、值values
* 缺失值要想办法按属性做填充会比较好看

## pivot_table 按总值排序
### 排序函数sort_values()
#### 参数说明
* by 	指定列名(axis=0或’index’)或索引值(axis=1或’columns’)
* axis 	若axis=0或’index’，则按照指定列中数据大小排序；若axis=1或’columns’，则按照指定索引中数据大小排序，默认axis=0
* ascending 	是否按指定列的数组升序排列，默认为True，即升序排列
* inplace 	是否用排序后的数据集替换原来的数据，默认为False，即不替换
* na_position 	{‘first’,‘last’}，设定缺失值的显示位置


In [5]:
排序 = 分进合击迸阶.sum(axis=1).sort_values(ascending=False).index
排序
排序_总值 = 分进合击迸阶.sum(axis=1).sort_values(ascending=False).values
排序_总值
分进合击迸阶_排序过 =  分进合击迸阶.loc[排序]
分进合击迸阶_排序过
分进合击迸阶_排序过['总值'] = 排序_总值
分进合击迸阶_排序过

城市           -  Burlington Massachussets  Emerville  Foster City  Guilford  \
行业                                                                           
金融科技       NaN                       NaN        NaN          NaN       NaN   
共享经济       NaN                       NaN        NaN          NaN       NaN   
电子商务       NaN                       NaN        NaN          NaN       NaN   
媒体和娱乐      NaN                       NaN        NaN          NaN       NaN   
云计算       70.0                       NaN        NaN          NaN       NaN   
物流         NaN                       NaN        NaN          NaN       NaN   
人工智能       NaN                       NaN        NaN        200.0       NaN   
消费品      150.0                       NaN        NaN          NaN       NaN   
生命科学       NaN                       NaN        NaN          NaN       NaN   
健康科技       NaN                       NaN        NaN          NaN      70.0   
区块链      150.0                       NaN        NaN          NaN       NaN   
航天         NaN                       NaN        NaN          NaN       NaN   
大数据        NaN                       NaN        NaN          NaN       NaN   
新能源汽车      NaN                       NaN        NaN          NaN       NaN   
软件与服务      NaN                       NaN        NaN          NaN       NaN   
房地产科技      NaN                       NaN        NaN          NaN       NaN   
教育科技       NaN                       NaN        NaN          NaN       NaN   
游戏         NaN                       NaN        NaN          NaN       NaN   
机器人        NaN                       NaN        NaN          NaN       NaN   
网络安全       NaN                       NaN      500.0          NaN       NaN   
新零售        NaN                       NaN        NaN          NaN       NaN   
新能源        NaN                       NaN        NaN          NaN       NaN   
虚拟与增强现实    NaN                       NaN        NaN          NaN       NaN   
即时通讯       NaN                       NaN        NaN          NaN       NaN   
3D印刷       NaN                     150.0        NaN          NaN       NaN   

城市       Harrisburg  Plantation  Stafford      上海     东京  ...    阿林顿    雅加达  \
行业                                                        ...                 
金融科技            NaN         NaN       NaN  3470.0    NaN  ...    NaN    NaN   
共享经济            NaN         NaN       NaN   200.0    NaN  ...    NaN  700.0   
电子商务            NaN         NaN       NaN   780.0    NaN  ...    NaN  870.0   
媒体和娱乐           NaN         NaN       NaN   590.0    NaN  ...    NaN    NaN   
云计算             NaN         NaN       NaN    70.0    NaN  ...    NaN    NaN   
物流              NaN         NaN       NaN   440.0    NaN  ...    NaN    NaN   
人工智能          500.0         NaN       NaN   290.0  150.0  ...    NaN    NaN   
消费品             NaN         NaN       NaN   100.0    NaN  ...    NaN    NaN   
生命科学            NaN         NaN       NaN   300.0    NaN  ...    NaN    NaN   
健康科技            NaN         NaN       NaN   970.0    NaN  ...    NaN    NaN   
区块链             NaN         NaN       NaN     NaN   70.0  ...    NaN    NaN   
航天              NaN         NaN       NaN     NaN    NaN  ...  200.0    NaN   
大数据             NaN         NaN       NaN    70.0    NaN  ...    NaN    NaN   
新能源汽车           NaN         NaN       NaN   600.0    NaN  ...    NaN    NaN   
软件与服务           NaN         NaN       NaN   420.0    NaN  ...    NaN    NaN   
房地产科技           NaN         NaN       NaN   170.0    NaN  ...    NaN    NaN   
教育科技            NaN         NaN       NaN   450.0    NaN  ...    NaN    NaN   
游戏              NaN         NaN       NaN     NaN    NaN  ...    NaN    NaN   
机器人             NaN         NaN       NaN     NaN    NaN  ...    NaN    NaN   
网络安全            NaN         NaN       NaN     NaN    NaN  ...    NaN    NaN   
新零售             NaN         NaN       NaN     NaN    NaN  ...    NaN    NaN   
新能源             NaN         NaN      70.0    70.0    NaN  ...    NaN    NaN 

# 小结
* groupby是一维的 表比较长、窄 机器喜欢看，pivot是二维的 表比较宽 人爱看
* 人的视觉思考一般在二维比较多 
* 思考同样的表格我要怎么做才能更容易让人理解
* 做数据分析数据可视化的时候要分析总结别人习惯，如时间轴通常我们的习惯是放在横轴从左到右